### PR TITLE
924 introduce default depth param

### LINKED ISSUE #924: range_max sensor value introduced but not used
I was wondering why in point_cloud_xyz.cpp in line 112 convertDepth is called but there is no way to set the max_sensor range that is part of the function based on a parameter given to the nodelet. I mean why implement the max sensor range variable and then not make it usable?

The option to set this parameter when calling convertDepth is not used in a single nodelet definition and would be a really useful tool for map generation. Perhaps someone can shed some light on that choice.

### PR DESCRIPTION
**Pull request by PhilippPolterauer**
this pull request attempts to fix #924 

changes:
- [x] changed the argument name `range_max` which was missleading to `invalid_depth` which captures the true meaning of the paramter
- [x] Introduction of a parameter called `invalid_depth`, which is the value that will be assigned to points that are not valid according to `DepthTraits<T>::valid`
- [x] small fixes regarding use of non const shared ptr references
- [x] changed for loop from int to uint32_t avoiding many type casts

Open:

- [ ] UB when reinterpret casting the pointcloud data? 
- [ ] introduction of a true "max_depth" or "max_range" paramter 

Comment:
i think the term range and depth is not always used consistently, my understanding is
*range* meaning point to point distance, and *depth* point to camera plane distance

### CODE REVIEW

File: .gitignore
Comment (mikeferguson) [inline .gitignore]: these folders shouldn't be inside your checked out repo - the usual layout is:

```
build
log
install
src
   this_repo
   other_repos
```

See https://docs.ros.org/en/rolling/Tutorials/Beginner-Client-Libraries/Creating-A-Workspace/Creating-A-Workspace.html

File: depth_image_proc/include/depth_image_proc/conversions.hpp
Comment (mikeferguson) [inline depth_image_proc/include/depth_image_proc/conversions.hpp:56]: I actually think range_max is a better name simply because that's more in line with what a number of depth camera drivers define (as well as most laser range finders). I'd suggest we keep the name of this - but add a doxygen comment to the function to explain what range_max is used for.

File: depth_image_proc/include/depth_image_proc/conversions.hpp
Comment (mikeferguson) [inline depth_image_proc/include/depth_image_proc/conversions.hpp]: this isn't undefined behavior - the PointCloud2Iterator is actually designed for exactly this use case - it computes what the offset is for each of the x/y/z elements inside the point cloud buffer

File: .gitignore
Comment (PhilippPolterauer) [reply .gitignore]: Agreed i will remove the changes in the gitigniore.

File: depth_image_proc/include/depth_image_proc/conversions.hpp
Comment (PhilippPolterauer) [reply depth_image_proc/include/depth_image_proc/conversions.hpp:56]: Did you check the implementation? If i Recall correctly ( i am currently writing from my phone) this parameter is only used for invalid points. So if there are Points with more than `range_max` distance (depth) they will not be clamped to this value, which ist contrary to what its name indicates IMHO

File: depth_image_proc/include/depth_image_proc/conversions.hpp
Comment (PhilippPolterauer) [inline depth_image_proc/include/depth_image_proc/conversions.hpp:87]: here you can see that the `range_max` values only use was to replace invalid data with this, but not limit other points.

File: depth_image_proc/include/depth_image_proc/conversions.hpp
Comment (PhilippPolterauer) [reply depth_image_proc/include/depth_image_proc/conversions.hpp]: >  const T * depth_row = reinterpret_cast<const T *>(&depth_msg->data[0]);

this is the line my comment is refering to.  I am uncertain about strict aliasing rules here. as there is no constraint on T guaranteeing this should be defined behavior. My suggestion would be to use std::copy which should in most trivial cases be optimized into a basic reinterpretation.

**Comment by PhilippPolterauer** [comment]
in the meantime i further research the UB topic, and after watching [Video](https://www.youtube.com/watch?v=_qzMpk-22cc) i am now fairly certain, that most use of reinterpret_cast is actually UB. Should i raise a new issue dedicated to this?

**Comment by mikeferguson** [comment]
Yeah, go ahead and open a new issue where we can further investigate the reinterpret_cast

File: depth_image_proc/include/depth_image_proc/conversions.hpp
Comment (mikeferguson) [reply depth_image_proc/include/depth_image_proc/conversions.hpp]: Let's remove the comment - but file a ticket to investigate

File: depth_image_proc/src/point_cloud_xyz.cpp
Comment (mikeferguson) [inline depth_image_proc/src/point_cloud_xyz.cpp:62]: for completeness, we should probably add this parameter to the xyzrgb and xyzi components. This new parameter also needs to be added to the documentation for each of the components - https://github.com/ros-perception/image_pipeline/blob/rolling/depth_image_proc/doc/components.rst

**Review by mikeferguson — CHANGES_REQUESTED** [review]
A couple of minor items - once we address those, this can be merged (I'm going to just double check things by firing up my robot on this branch tomorrow to make sure the default behavior hasn't changed even though it all looks correct)

**Comment by PhilippPolterauer** [comment]
Thanks for the feedback. As soon as i have time, i will implement the changes.

File: depth_image_proc/include/depth_image_proc/conversions.hpp
Comment (PhilippPolterauer) [reply depth_image_proc/include/depth_image_proc/conversions.hpp]:
```suggestion
```

**Comment by PhilippPolterauer** [comment]
- [x] added `invalid_depth` to parameter description in the docs
- [x] removed comment
- [x] added the `invalid_depth` to the other  non radial nodes as well.

If there is something more to do, please let me know!

**Review by mikeferguson — APPROVED** [review]

**Comment by mikeferguson** [comment]
This looks good - thanks for the contribution - just waiting on the final CI to pass

**Comment by PhilippPolterauer** [comment]
i just clicked `update branch` which i think was a mistake, since now the ci is waiting again. sorry for the confusion :)

**Pull request merged**
